### PR TITLE
OperatorHub Integration - WIP

### PR DESCRIPTION
**Pull request by raffaelespazzoli**
## Description
this PR prepares the operator to be made available in operator hub when a release is done.
the readme contains the steps to be followed when releasing. For now the steps are manual. 

Please follow the steps until you see what this operator would look like in operator hub and give some feedback there.

Once satisfied I'll test the installation from operator hub actually works and then at the following release we can also push this to operator hub.

because of how the `operator-sdk` and `operator-courier` commands work I had to create some manifest files that are actually duplicates of the manifest that are in the helm charts. May in the future we can think of more sophisticated solutions.

### CODE REVIEW

File: deploy/olm-catalog/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
+# Operator Hub Release steps
```
Comment (seanmalloy) [inline deploy/olm-catalog/README.md]: I think we should move these to a new file named `.github/RELEASE.md`. There are some additional release documentation that we can add in the future.

**Comment by akavel** [comment]
There seem to be many files very similar (identical?) to ones already in the repo (e.g. the CRD); is this PR intending to add duplicated files? can't we just prepare a temporary dir when making a release to OperatorHub? is it a requirement of OperatorHub that the repo has specific files at specific locations?

File: deploy/olm-catalog/README.md (hunk 1 above)
Comment (rhalat) [reply deploy/olm-catalog/README.md]: Done @seanmalloy

**Comment by rhalat** [comment]
@akavel `deploy/olm-catalog/eunomia` directory needs to be submitted to [community-operators repo](https://github.com/operator-framework/community-operators/tree/master/community-operators) and I believe we don't need it here, but in this PR it's for code review purposes.

**Comment by seanmalloy** [comment]
@rhalat please go ahead and submit the pull request to https://github.com/operator-framework/community-operators/tree/master/community-operators. Please @ mention me on the pull request.

I believe we will want to still merge the file `.github/RELEASE.md` file in this pull request.

**Comment by rhalat** [comment]
@seanmalloy I tried to follow @raffaelespazzoli instructions and I managed to add eunomia v0.0.2 to my private quay application registry https://quay.io/application/zezima/eunomia?tab=releases 

I also created temporary PR from eunomia-0.0.2 to master branch in my fork here https://github.com/rhalat/community-operators/pull/1 but as for now I didn't test deployment, because I don't know exactly how. In this case I'm not sure if I should submit PR to https://github.com/operator-framework/community-operators/tree/master/community-operators now or first try to test it somehow.

**Comment by seanmalloy** [comment]
@rhalat did you see this documentation ... https://operatorhub.io/contribute. This document has a link to https://github.com/operator-framework/community-operators/blob/master/docs/testing-operators.md#manual-testing-on-kubernetes. It looks like the second link provides instructions for testing with minikube and OLM.

Also, why did you need to create a personal quay.io? Can't you just use https://quay.io/repository/kohlstechnology/eunomia-operator?

**Comment by akavel** [comment]
One more question has occurred to me: in future, do we expect to need to redo this process? — e.g. when the CRD changes, like with #134 — or not?

If yes, wouldn't it make sense to try and put as much of this process as possible in scripts, to make our life in future easier? I would imagine the scripts could e.g. build a temporary directory, with structure as expected for copying into community-operators; do you think this makes any sense, or not really?

**Comment by rhalat** [comment]
@seanmalloy Thank you for provided link. I did all the steps to test it in minikube, but operator installation failed with following events:

```
Events:
  Type     Reason               Age                From                        Message
  ----     ------               ----               ----                        -------
  Normal   RequirementsUnknown  54m (x2 over 54m)  operator-lifecycle-manager  requirements not yet checked
  Normal   RequirementsNotMet   54m (x2 over 54m)  operator-lifecycle-manager  one or more requirements couldn't be found
  Warning  InstallCheckFailed   49m (x2 over 49m)  operator-lifecycle-manager  install timeout
  Normal   NeedsReinstall       49m (x2 over 49m)  operator-lifecycle-manager  installing: Waiting: waiting for deployment eunomia-operator to become ready: Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
  Normal   AllRequirementsMet   49m (x4 over 54m)  operator-lifecycle-manager  all requirements found, attempting install
  Normal   InstallSucceeded     49m (x3 over 54m)  operator-lifecycle-manager  waiting for install components to report healthy
  Normal   InstallWaiting       49m (x4 over 54m)  operator-lifecycle-manager  installing: Waiting: waiting for deployment eunomia-operator to become ready: Waiting for rollout to finish: 0 out of 1 new replicas have been updated...
  Warning  InstallCheckFailed   44m (x2 over 44m)  operator-lifecycle-manager  install failed: Timeout: deployment eunomia-operator not ready before timeout: deployment "eunomia-operator" exceeded its progress deadline
```

It was required for testing to have eunomia in application registry. In kohlstechnology quay.io account we have it only in container registry, that's why I created my private one to push it there.

**Comment by rhalat** [comment]
Looks like it's working after I added all the resources from `deploy` directory manually.

```
$ kubectl get clusterserviceversion -n marketplace
NAME             DISPLAY   VERSION   REPLACES         PHASE
eunomia.v0.0.2   Eunomia   0.0.2     eunomia.v0.0.1   Succeeded
```

I also downgraded to kubernetes 1.15.0 from 1.16.2, but I'm not sure if this was required. Will do more testing.

**Comment by seanmalloy** [comment]
@akavel yes I believe we will need to do this with every time we release a new version. And yes in the future we will want to have as much automation as possible around this process, but for now we just want to see if we can get the first version listed on operatorhub.io.

**Comment by sbar95** [comment]
@seanmalloy I created a PR: https://github.com/operator-framework/community-operators/pull/902
I mentioned you and @raffaelespazzoli in a comment there.

To see how exactly eunomia-operator will look like in operatorhub, please paste the contents of [eunomia.v0.0.3.clusterserviceversion.yaml](https://github.com/KohlsTechnology/eunomia/blob/9d49b4a6c5be005de8315a230e2afab6bd1585a9/deploy/olm-catalog/eunomia/0.0.3/eunomia.v0.0.3.clusterserviceversion.yaml) to [operatorhub preview](https://operatorhub.io/preview).

I tested eunomia-operator locally as described in [manual tests](https://github.com/operator-framework/community-operators/blob/master/docs/testing-operators.md#manual-testing-on-kubernetes).
I tested watching:
- OwnNamespace (in this case both eunomia-operator and hello-world have been deployed to the marketplace namespace),
- SingleNamespace (here eunomia-operator has been deployed to the marketplace namespace, and hello-world to the eunomia-hello-world-yaml-demo namespace),
- AllNamespaces (eunomia-operator as above, hello-world to hello-world-yaml-demo and hello-world-yaml-demo2 namespaces).

Here is a screenshot from the last case:
image

**Comment by sbar95** [comment]
[CVP/operator-olm-deployment check](https://s3.amazonaws.com/cvp-co-pipeline-artifacts/cvp-community-operator-metadata-validation-test/902/7974ff8f-934f-4465-a6cd-c85d8d8f60b7/operator-olm-deployment-output.txt) is failing with the message:
`install failed: Timeout: deployment eunomia-operator not ready before timeout: deployment "eunomia-operator" exceeded its progress deadline`
That's because a manual step mentioned in my previous comment is required. Automatic check cannot perform it.

File: .github/RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
+# Operator Hub Release steps
+
+set some environment variables used during the process
```
[the remark this answers is not in the document]
Comment (sbar95) [reply .github/RELEASE.md]: In this PR Sean wanted to just [add eunomia to operatorhub.io](https://github.com/KohlsTechnology/eunomia/pull/86#issuecomment-550377577) but I can try to sort it out

[the remark this answers is not in the document]
Comment (seanmalloy) [reply .github/RELEASE.md]: We automate it once we have the initial versions of eunomia added to operatorhub.io.

File: deploy/configmap.yaml [hunk 3]
```diff
@@ -0,0 +1,97 @@
+apiVersion: v1
```
[the remark this answers is not in the document]
Comment (seanmalloy) [reply deploy/configmap.yaml]: See discussion at operator-framework/community-operators/pull/902. It looks like we need the operator to automatically crate this `ConfigMap` for the automated OLM deployment to work.

I created #162 for tracking this feature enhancement.

**Comment by sbar95** [comment]
@seanmalloy #162 needs to be completed before eunomia is made available on operatorhub either way, but I now see [the PR](https://github.com/operator-framework/community-operators/pull/902) I created and @rhalat's [temporary PR](https://github.com/rhalat/community-operators/pull/1) are wrong - [where-to-contribute](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#where-to-contribute). We should be putting eunomia in upstream-community-operators directory, not community-operators.

**Comment by seanmalloy** [comment]
@sbar95 I did not fully understand all of the steps to get an operator added to operatorhub.io when we started down this path. Thanks for your hard work on this!

Looking at the [ where-to-contribute](https://github.com/operator-framework/community-operators/blob/master/docs/contributing.md#where-to-contribute) doc that you linked to I think we will want eunomia added to both upstream-community-operators and community-operators. We can start with just getting it added to `community-operators` for now.

I'll get #162 added to the v0.0.5 milestone.

**Comment by rajneesh-ops** [comment]
Closing this PR. Created a new one
https://github.com/KohlsTechnology/eunomia/pull/235

**Pull request closed without merging**